lib/bashcov/runner.rb

```diff
@@ -37,48 +64,96 @@ def result
         expunge_invalid_files!
         mark_relevant_lines!
 
-        @coverage
+        convert_coverage
       end
     end
 
   private
 
+    def write_warning(message)
+      warn format "%s: warning: %s", Bashcov.program_name,
+                  message.gsub(/^\s+/, "").lines.map(&:chomp).join(" ")
+    end
+
+    def run_xtrace(fd, env, options)
+      # Older versions of Bash (< 4.1) don't have the BASH_XTRACEFD variable
+      if Bashcov.bash_xtracefd?
+        options[fd] = fd # bind FDs to the child process
+
+        env["BASH_XTRACEFD"] = fd.to_s
+      else
+        # Send subprocess standard error to @xtrace.file_descriptor
+        options[:err] = fd
+
+        # Don't bother issuing warning if we're silencing output anyway
+        unless Bashcov.mute
+          write_warning <<-WARNING
+            you are using a version of Bash that does not support
+            BASH_XTRACEFD. All xtrace output will print to standard error, and
+            your script's output on standard error will not be printed to the
+            console.
+          WARNING
+        end
+      end
+
+      inject_xtrace_flag! do
+        yield
+      end
+    end
+
     # @note +SHELLOPTS+ must be exported so we use Ruby's {ENV} variable
-    # @return [void]
+    # @yield [void] adds "xtrace" to +SHELLOPTS+ and then runs the provided
+    #   block
+    # @return [Object, ...] the value returned by the calling block
     def inject_xtrace_flag!
-      existing_flags = (ENV["SHELLOPTS"] || "").split(":")
+      return enum_for(__method__) unless block_given?
+
+      existing_flags_s = ENV["SHELLOPTS"]
+      existing_flags = (existing_flags_s || "").split(":")
       ENV["SHELLOPTS"] = (existing_flags | ["xtrace"]).join(":")
+
+      # Calls the provided block
+      Proc.new.call
```

Comment:
Sweet, are you up for fixing these @BaxterStockman or should I merge this PR and do it myself? I'm fine either way, just let me know :)